06_puppy/kasunfdo/types.go

```diff
@@ -0,0 +1,29 @@
+package main
+
+import (
+	"sync"
+)
+
+// Data structure for puppy
+type Puppy struct {
+	ID     uint64
+	Breed  string
+	Colour string
+	Value  float64
+}
+
+// Storer defines standard CRUD operations for Puppy
+type Storer interface {
+	CreatePuppy(puppy *Puppy) bool
+	ReadPuppy(id uint64) (*Puppy, bool)
+	UpdatePuppy(id uint64, puppy *Puppy) bool
+	DeletePuppy(id uint64) bool
+}
+
+// MapStore type and implementation
+type MapStore map[uint64]*Puppy
```

Comment:
I'm unsure (especially as I'm still a golang newbie) if this would be better as a value with type ```Puppy``` or ```*Puppy```.

Downside of having as a pointer, means, you store something in the Store, and then update that variable you passed to CreatePuppy, and what is stored in the Store has changed.

I'm unsure if this is a problem or not, if you feel confident in this design choice, feel free to make a comment, and resolve. It might something the higher-level reviewers might have an opinion on.

Reply:
Thank you for pointing out the potential issue. Working with pointers should be done with extra care. Both pointer and value implementations have their on pros and cons as well. I'm open for a discussion about that.

Reply:
Until requirement is clear, can't say either `Puppy` or `*Puppy` is right or wrong.

Reply:
@lexual is correct that with this implementation, you can alter what is stored without going through the storer interface, but that is not a given due to this type. The implementation could make a copy and still store pointers, and also return a copy. This may make sense because a typical implementation of a "storer" would be to store in an external database, in which case you could not alter the underlying storage through the pointer. But that more describes the interface than this `MapStore` implementation type.

I think it the most sensible approach is to use pointers in the interface and store values in the map.

@alfredxiao There are no requirements on this. It is an implementation detail. But as usual, any assumptions made should be documented.